Enable status button if current theme is not active.

mohdsayed opened the pull request:
Fixes https://github.com/xwp/wp-customize-snapshots/issues/131

mohdsayed committed:
Enable status button if current theme is not active.

mohdsayed commented:
@westonruter We have a param `theme=twentyfifteen` in url when we switch theme, I think that should also be saved in changeset like we saved the history query params in the other PR?

westonruter commented:
@sayedtaqui oh yes, absolutely! That param can be included in the customizer state query params only when saving a changeset for a theme that is not active. If the theme is active, then the param should be omitted from being saved.

mohdsayed committed:
Ensure that api.previewer.query is extended properly and only once.

mohdsayed committed:
Save previewing theme with customizer state  query params.

mohdsayed committed:
Enable update button for 4.6.x when theme is not active

mohdsayed committed:
Add current theme to preview link and enable save button when you switch back to the original theme.

mohdsayed committed:
Enable status button when ajax request fails.

westonruter commented:
Is this out of WIP and ready for review?

westonruter commented on js/customize-snapshots.js:
```diff
@@ -104,15 +107,22 @@
 			queryVars.url = api.previewer.previewUrl.get();
 
+			if ( ! api.state( 'activated' ).get() || snapshot.data.theme !== currentTheme ) {
+				queryVars.previewingTheme = true;
```
Since this is going to be sent with the POST request, I think it should be `previewing_theme` instead. Then below in PHP, you'd read ` $query_vars['previewing_theme']`.

westonruter commented on php/class-post-type.php:
```diff
@@ -917,11 +920,17 @@ public function get_frontend_view_link( $post ) {
+
+		if ( isset( $preview_url_query_vars['theme'] ) && $current_theme !== $preview_url_query_vars['theme'] ) {
+			$args = array_merge( $args, array(
+					'theme' => $preview_url_query_vars['theme'],
```
Extra tab here?

mohdsayed commented:
Not ready yet, need to test and may be few changes. Trying to finish today.

mohdsayed committed:
Improve disable status button logic and review feedback.

mohdsayed committed:
Fix condition for enabling status button when changing themes.

mohdsayed committed:
Update save button disable condition for 4.6.x when theme is changed.

mohdsayed commented:
Ready for review.

westonruter committed:
Ensure snapshots can load for previewing theme switch; re-use is_theme_active

westonruter committed:
Fix casting of booleans and numbers to strings in PHP exports to JS

westonruter commented on js/customize-snapshots.js:
```diff
@@ -255,6 +266,7 @@
 					dialogElement;
 
 				snapshot.statusButton.disableSelect.set( false );
+				snapshot.statusButton.disbleButton.set( false );
```
I'm glad you caught this. It's something I was noticing in the case of setting validation errors, where I'd get a validation error message with a control but the Publish button would remain disabled.

westonruter committed:
Ensure frontend preview link gets customize_theme param in theme switches

The 'customize_theme' param is used on the frontend, whereas just 'theme' is used in the admin.

westonruter commented:
@sayedtaqui the one remaining issue I see, and this isn't strictly related to the changes here, but when I publish the changeset the frontend preview link continues to have the `customize_changeset_uuid` even though it is published:

![image](https://cloud.githubusercontent.com/assets/134745/23984604/95bcb95c-09d7-11e7-8ecd-afca0f686584.png)

mohdsayed commented on php/class-customize-snapshot-manager-back-compat.php:
```diff
@@ -106,7 +101,7 @@ public function enqueue_controls_scripts() {
 			'initialServerTimestamp' => floor( microtime( true ) * 1000 ),
 			'theme' => $this->original_stylesheet,
-			'previewingTheme' => isset( $preview_url_query_vars['theme'] ) ? $preview_url_query_vars['theme']: '',
+			'previewingTheme' => ! $this->customize_manager->is_theme_active(),
```
@westonruter This is breaking my condition here https://github.com/xwp/wp-customize-snapshots/blob/9ca3ff2f90df27da5fc62946b7857b61dbcc3184/js/compat/customize-snapshots.js#L278
There are several use cases where the button has to be enabled or disabled.

1.  When they switch theme ( Has to enable ) ( Working )
2. After saving the previewing theme, reload the page, ( Has to disable ) ( Not working now )
3. After saving the previewing theme they switch to a third theme ( Has to enable ) ( Working )
4. After travelling through the themes and saving the last one, they come back to the original theme ( Should be enable ) ( Not working now ).

I was not able to satisfy these conditions with only boolean value of `previewingTheme `. This is as good as `api.state( 'activated' ).get()`

westonruter replied:
Thanks for pointing this out. So this can be fixed by reverting back to:

```php
'previewingTheme' => isset( $preview_url_query_vars['theme'] ) ? $preview_url_query_vars['theme'] : '',
```
?

mohdsayed replied:
Yes.

mohdsayed committed:
Remove uuid from preview link after publish

mohdsayed committed:
Revert the previewingTheme value.

mohdsayed commented on php/class-customize-snapshot-manager.php:
```diff
@@ -225,7 +226,7 @@ public function add_snapshot_uuid_to_return_url() {
 			&&
-			$this->is_theme_active()
+			$this->customize_manager->is_theme_active()
 			&&
```
@westonruter Not sure why `$this->customize_manager->is_theme_active()` is returning `false` in 4.6.1 which is failing phpunit test.  I mean inside phpunit test.

westonruter replied:
It's because the `setup_theme` action wasn't triggered in the test case, and so `\WP_Customize_Manager::$original_stylesheet` didn't get set. In 4.7 this variable gets set in `\WP_Customize_Manager::__construct()` whereas in 4.6 it got set in `\WP_Customize_Manager::setup_theme()`.

westonruter replied:
Should be fixed in 52694b693.

westonruter committed:
Ensure setup_theme is triggered in test so that WP_Customize_Manager::$original_stylesheet will be set
